File: backend/models/trade_group.go
package models

import (
	"errors"
	"time"

	"github.com/cloudmanic/app.options.cafe/backend/library/services"
)

type TradeGroup struct {
	Id               uint       `gorm:"primary_key" json:"id"`
	UserId           uint       `sql:"not null;index:UserId" json:"_"`
	CreatedAt        time.Time  `json:"_"`
	UpdatedAt        time.Time  `json:"_"`
	Name             string     `json:"name"`
	BrokerAccountId  uint       `sql:"not null;index:BrokerAccountId" json:"_"`
	BrokerAccountRef string     `sql:"not null;index:BrokerAccountRef" json:"broker_account_ref"`
	Status           string     `sql:"not null;type:ENUM('Open', 'Closed');default:'Open'" json:"status"`
	Type             string     `sql:"not null;type:ENUM('Option', 'Stock', 'Put Credit Spread', 'Call Credit Spread', 'Put Debit Spread', 'Call Debit Spread', 'Iron Condor', 'Other'); default:'Other'" json:"type"`
	OrderIds         string     `json:"_"`
	Risked           float64    `sql:"type:DECIMAL(12,2)" json:"risked"`
	Credit           float64    `sql:"type:DECIMAL(12,2)" json:"credit"`       // Before Commission
	Proceeds         float64    `sql:"type:DECIMAL(12,2)" json:"proceeds"`     // Before Commission
	Profit           float64    `sql:"type:DECIMAL(12,2)" json:"profit"`       // After Commission
	PercentGain      float64    `sql:"type:DECIMAL(12,2)" json:"percent_gain"` // After Commission
	Commission       float64    `sql:"type:DECIMAL(12,2)" json:"commission"`
	Note             string     `sql:"type:text" json:"note"`
	Positions        []Position `json:"positions"`
	OpenDate         time.Time  `json:"open_date"`
	ClosedDate       time.Time  `json:"closed_date"`
}

//
// List different trade groups.
//
func (t *DB) GetTradeGroups(params QueryParam) ([]TradeGroup, QueryMetaData, error) {

	var results = []TradeGroup{}

	// Run the query
	noFilterCount, err := t.QueryWithNoFilterCount(&results, params)

	// Throw error if we have one
	if err != nil {
		return results, QueryMetaData{}, err
	}

	// Get the meta data related to this query.
	meta := t.GetQueryMetaData(len(results), noFilterCount, params)

	// Add the symbol data to the positions.
	t.tradeGroupAddSymbolsToPositions(results)

	// Return happy
	return results, meta, nil
}

//
// Get TradeGroup by Id
//
func (t *DB) GetTradeGroupById(id uint) (TradeGroup, error) {

	tg := TradeGroup{}

	if t.Preload("Positions").Where("Id = ?", id).First(&tg).RecordNotFound() {
		return tg, errors.New("Record not found")
	}

	// Loop through and add the symbol to the positions object
	for key, row := range tg.Positions {
		t.Model(row).Related(&tg.Positions[key].Symbol)
	}

	// Return happy
	return tg, nil
}

//
// Store a new TradeGroup.
//
func (t *DB) CreateTradeGroup(tg *TradeGroup) error {

	// Create order
	t.Create(tg)

	// Return happy
	return nil
}

//
// Update a TradeGroup.
//
func (t *DB) UpdateTradeGroup(tg *TradeGroup) error {

	// Update entry.
	t.Save(&tg)

	// Return happy
	return nil
}

//
// Helpful function for adding symbols to positions
//
func (t *DB) tradeGroupAddSymbolsToPositions(tgs []TradeGroup) error {

	// Loop through and add the symbol to the positions object
	for key, row := range tgs {
		for key2, row2 := range tgs[key].Positions {
			t.Model(row2).Related(&tgs[key].Positions[key2].Symbol)
		}

		// Stupid code to make it so GoFMT does not delete in range statement
		if row.Id < 0 {
			services.Info("This should never happen.")
		}
	}

	return nil
}

/* End File */

File: frontend/src/app/models/trade-group.ts
//
// Date: 2/10/2018
// Author(s): Spicer Matthews (spicer@options.cafe)
// Copyright: 2018 Cloudmanic Labs, LLC. All rights reserved.
//

import { Symbol } from './symbol';
import { Position } from './position';

//
// TradeGroup Model
//
export class TradeGroup 
{
  public Id: number;
  public Name: string;
  public Status: string;
  public Type: string;
  public OpenDate: string;
  public ClosedDate: string;
  public Credit: number;
  public Profit: number;
  public Proceeds: number;  
  public PercentGain: number;
  public Risked: number;
  public Commission: number;
  public Note: string;
  public Positions: Position[];

  //
  // Constructor
  //
  constructor(
    Id: number, 
    Name: string,
    Status: string, 
    Type: string,
    OpenDate: string,
    ClosedDate: string,
    Credit: number,
    Profit: number,
    Proceeds: number,    
    PercentGain: number,
    Risked: number,
    Commission: number,
    Note: string,
    Positions: Position[]        
  ) {
    this.Id = Id;
    this.Name = Name;
    this.Status = Status;
    this.Type = Type;
    this.OpenDate = OpenDate;
    this.ClosedDate = ClosedDate;
    this.Credit = Credit;
    this.Profit = Profit;
    this.Proceeds = Proceeds;    
    this.PercentGain = PercentGain;
    this.Risked = Risked;
    this.Commission = Commission;
    this.Note = Note;
    this.Positions = Positions;
  }

  //
  // Build object for emitting to the app.
  //
  public static buildForEmit(data) : TradeGroup[]  
  {

    let tg = [];

    if(! data)
    {
      return tg;      
    }

    for(let i = 0; i < data.length; i++)
    {
      // Add in the positions
      let positions = [];
      
      for(let k = 0; k < data[i].positions.length; k++)
      {
        positions.push(new Position(
          data[i].positions[k].id,
          data[i].positions[k].open_date, 
          data[i].positions[k].close_date, 
          data[i].positions[k].qty,
          data[i].positions[k].org_qty,
          data[i].positions[k].cost_basis,
          data[i].positions[k].proceeds,
          data[i].positions[k].profit,
          new Symbol().fromJson(data[i].positions[k].symbol)                           
        ));
      }

      tg.push(new TradeGroup(
        data[i].id, 
        data[i].name,         
        data[i].status, 
        data[i].type, 
        data[i].open_date, 
        data[i].closed_date,
        data[i].credit,
        data[i].profit,
        data[i].proceeds,
        data[i].percent_gain,
        data[i].risked,
        data[i].commission,
        data[i].note,
        positions
       ));
    }

    return tg; 
  }
}

//
// Setup a class to hold all the different position types
//
export class TradeGroupsCont 
{
  public Equity: TradeGroup[] = []; 
  public ShortEquity: TradeGroup[] = []; 
  public Option: TradeGroup[] = []; 
  public PutCreditSpread: TradeGroup[] = []; 
  public CallCreditSpread: TradeGroup[] = []; 
  public PutDebitSpread: TradeGroup[] = []; 
  public CallDebitSpread: TradeGroup[] = []; 
  public IronCondor: TradeGroup[] = []; 
  public ReverseIronCondor: TradeGroup[] = [];
  public LongCallButterfly: TradeGroup[] = [];
  public LongPutButterfly: TradeGroup[] = [];   
  public Other: TradeGroup[] = []; 
}

/* End Find */